RemoteLayerBackingStoreCollection doesn't support mixing remote rendering with in process rendering.

mattwoodrow opened the pull request:
#### 12198a179de6bbb9c9e7da80441dfe77a3d71c27
<pre>
RemoteLayerBackingStoreCollection doesn&apos;t support mixing remote rendering with in process rendering.
<a href="https://bugs.webkit.org/show_bug.cgi?id=265617">https://bugs.webkit.org/show_bug.cgi?id=265617</a>
&lt;<a href="https://rdar.apple.com/119010082">rdar://119010082</a>&gt;

Reviewed by Tim Horton.

We currently have a polymorphic RemoteLayerBackingStoreCollection object, which
 requires an upfront decision about whether in-process or remote rendering will
be used.

We should instead have a single RemoteLayerBackingStoreCollection which supports
mixing in-process and remote RemoteLayerBackingStore objects as needed.

This merges RemoteLayerBackingStoreCollection with the &apos;WithRemoteRendering&apos; variant,
and the resulting code checks the type of each backing stores and handles them
separately (for functions that require it).

RemoteLayerBackingStore is split to have two subclasses (&apos;Remote&apos; and &apos;InProcess&apos;),
and code that only applies to that situation is moved out of the base class.

We still currently only create a single type of backing store based on the global
flag, but mixing types should now be supported.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::isRemoteLayerWithRemoteRenderingBackingStore const):
(WebKit::RemoteLayerBackingStore::isRemoteLayerWithInProcessRenderingBackingStore const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerBackingStore::swapToValidFrontBuffer): Deleted.
(WebKit::RemoteLayerBackingStore::applySwappedBuffers): Deleted.
(WebKit::RemoteLayerBackingStore::prepareToDisplay): Deleted.
(WebKit::RemoteLayerBackingStore::prepareBuffers): Deleted.
(WebKit::RemoteLayerBackingStore::setBufferVolatile): Deleted.
(WebKit::RemoteLayerBackingStore::setBufferNonVolatile): Deleted.
(WebKit::RemoteLayerBackingStore::setFrontBufferNonVolatile): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
(WebKit::RemoteLayerBackingStoreCollection::layerTreeContext const):
(WebKit::RemoteLayerBackingStoreCollection::markInProcessBackingStoreVolatile):
(WebKit::RemoteLayerBackingStoreCollection::gpuProcessConnectionWasDestroyed): Deleted.
(WebKit::RemoteLayerBackingStoreCollection::markBackingStoreVolatile): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::RemoteLayerBackingStoreCollection):
(WebKit::RemoteLayerBackingStoreCollection::prepareBackingStoresForDisplay):
(WebKit::RemoteLayerBackingStoreCollection::createRemoteLayerBackingStore):
(WebKit::RemoteLayerBackingStoreCollection::markInProcessBackingStoreVolatile):
(WebKit::RemoteLayerBackingStoreCollection::markBackingStoreVolatileAfterReachabilityChange):
(WebKit::RemoteLayerBackingStoreCollection::markAllBackingStoreVolatile):
(WebKit::RemoteLayerBackingStoreCollection::tryMarkAllBackingStoreVolatile):
(WebKit::RemoteLayerBackingStoreCollection::markAllBackingStoreVolatileFromTimer):
(WebKit::RemoteLayerBackingStoreCollection::gpuProcessConnectionWasDestroyed):
(WebKit::RemoteLayerBackingStoreCollection::collectRemoteRenderingBackingStoreBufferIdentifiersToMarkVolatile):
(WebKit::RemoteLayerBackingStoreCollection::collectAllRemoteRenderingBufferIdentifiersToMarkVolatile):
(WebKit::RemoteLayerBackingStoreCollection::sendMarkBuffersVolatile):
(WebKit::RemoteLayerBackingStoreCollection::markBackingStoreVolatile): Deleted.
(WebKit::RemoteLayerBackingStoreCollection::allocateBufferForBackingStore): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h: Added.
(isType):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm: Added.
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::prepareBuffers):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::setBufferVolatile):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::setBufferNonVolatile):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::setFrontBufferNonVolatile):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::prepareToDisplay):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::swapToValidFrontBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.h: Added.
(isType):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm: Added.
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::applySwappedBuffers):
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::allocateBuffer const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.h: Removed.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.mm: Removed.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::ensureBackingStore):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::RemoteLayerTreeContext):

Canonical link: <a href="https://commits.webkit.org/271529@main">https://commits.webkit.org/271529@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66a0487d550b600477c3f905cfb54b46413545b2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28651 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30024 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31278 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26151 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9408 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4657 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28921 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6026 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24627 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5239 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5401 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32614 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26226 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26074 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31632 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5356 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3527 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29409 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6968 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6862 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5822 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5874 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
```diff
@@ -77,6 +77,9 @@ class RemoteLayerBackingStore : public CanMakeWeakPtr<RemoteLayerBackingStore> {
     enum class IncludeDisplayList : bool { No, Yes };
 #endif
 
+    virtual bool isRemoteLayerWithRemoteRenderingBackingStore() const { return false; }
```
I'd love to remove the two different meanings of "remote" here.

mattwoodrow committed:
RemoteLayerBackingStoreCollection doesn't support mixing remote rendering with in process rendering.
https://bugs.webkit.org/show_bug.cgi?id=265617
<rdar://119010082>

Reviewed by Tim Horton.

We currently have a polymorphic RemoteLayerBackingStoreCollection object, which
 requires an upfront decision about whether in-process or remote rendering will
be used.

We should instead have a single RemoteLayerBackingStoreCollection which supports
mixing in-process and remote RemoteLayerBackingStore objects as needed.

This merges RemoteLayerBackingStoreCollection with the 'WithRemoteRendering' variant,
and the resulting code checks the type of each backing stores and handles them
separately (for functions that require it).

RemoteLayerBackingStore is split to have two subclasses ('Remote' and 'InProcess'),
and code that only applies to that situation is moved out of the base class.

We still currently only create a single type of backing store based on the global
flag, but mixing types should now be supported.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::isRemoteLayerWithRemoteRenderingBackingStore const):
(WebKit::RemoteLayerBackingStore::isRemoteLayerWithInProcessRenderingBackingStore const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerBackingStore::swapToValidFrontBuffer): Deleted.
(WebKit::RemoteLayerBackingStore::applySwappedBuffers): Deleted.
(WebKit::RemoteLayerBackingStore::prepareToDisplay): Deleted.
(WebKit::RemoteLayerBackingStore::prepareBuffers): Deleted.
(WebKit::RemoteLayerBackingStore::setBufferVolatile): Deleted.
(WebKit::RemoteLayerBackingStore::setBufferNonVolatile): Deleted.
(WebKit::RemoteLayerBackingStore::setFrontBufferNonVolatile): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
(WebKit::RemoteLayerBackingStoreCollection::layerTreeContext const):
(WebKit::RemoteLayerBackingStoreCollection::markInProcessBackingStoreVolatile):
(WebKit::RemoteLayerBackingStoreCollection::gpuProcessConnectionWasDestroyed): Deleted.
(WebKit::RemoteLayerBackingStoreCollection::markBackingStoreVolatile): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::RemoteLayerBackingStoreCollection):
(WebKit::RemoteLayerBackingStoreCollection::prepareBackingStoresForDisplay):
(WebKit::RemoteLayerBackingStoreCollection::createRemoteLayerBackingStore):
(WebKit::RemoteLayerBackingStoreCollection::markInProcessBackingStoreVolatile):
(WebKit::RemoteLayerBackingStoreCollection::markBackingStoreVolatileAfterReachabilityChange):
(WebKit::RemoteLayerBackingStoreCollection::markAllBackingStoreVolatile):
(WebKit::RemoteLayerBackingStoreCollection::tryMarkAllBackingStoreVolatile):
(WebKit::RemoteLayerBackingStoreCollection::markAllBackingStoreVolatileFromTimer):
(WebKit::RemoteLayerBackingStoreCollection::gpuProcessConnectionWasDestroyed):
(WebKit::RemoteLayerBackingStoreCollection::collectRemoteRenderingBackingStoreBufferIdentifiersToMarkVolatile):
(WebKit::RemoteLayerBackingStoreCollection::collectAllRemoteRenderingBufferIdentifiersToMarkVolatile):
(WebKit::RemoteLayerBackingStoreCollection::sendMarkBuffersVolatile):
(WebKit::RemoteLayerBackingStoreCollection::markBackingStoreVolatile): Deleted.
(WebKit::RemoteLayerBackingStoreCollection::allocateBufferForBackingStore): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h: Added.
(isType):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm: Added.
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::prepareBuffers):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::setBufferVolatile):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::setBufferNonVolatile):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::setFrontBufferNonVolatile):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::prepareToDisplay):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::swapToValidFrontBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.h: Added.
(isType):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm: Added.
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::applySwappedBuffers):
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::allocateBuffer const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.h: Removed.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.mm: Removed.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::ensureBackingStore):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::RemoteLayerTreeContext):

Canonical link: https://commits.webkit.org/271529@main
